sdk/tools/whitelister/whitelister.cpp

```diff
@@ -0,0 +1,301 @@
+    fprintf(stderr, "!! size of whitelist entry = %Iu\n", wlEntry.Lines.size());
+    Whitelist.push_back(wlEntry);
+}
+
+vector<WL_ENTRY>
+ReadWhitelist(string& WhitelistFileName)
+{
+    vector<WL_ENTRY> whitelist;
+
+    ifstream fileStream(WhitelistFileName);
+    for (string line; getline(fileStream, line); )
+    {
+        fprintf(stderr, "adding\n");
+        AddWhitelistEntry(whitelist, line);
+    }
+
+    return whitelist;
+}
+
+string
+CreateLinesTableCode(WL_ENTRY& WlEntry)
+{
+    string code;
+
+    // Get file name without extension
+    size_t nameLength = WlEntry.ShortFileName.rfind('.');
+    string sourceFileNameNoExt = WlEntry.ShortFileName.substr(0, nameLength);
+
+    code = "static WHITELIST_ENTRY s_Whitelist_" + sourceFileNameNoExt + "[] =\n";
+    code += "{\n";
+    for (auto& lineEntry : WlEntry.Lines)
+    {
+        code += "    { ";
+        code += to_string(lineEntry.Line);
+        code += ", \"" + lineEntry.Reason + "\"},\n";
+    }
+    code += "};\n\n";
+
+    return code;
+}
+
+string
+GenerateWhitelistCode(vector<WL_ENTRY>& Whitelist)
+{
+    string code;
+
+    fprintf(stderr, "size of whitelist = %Iu\n", Whitelist.size());
+
+    code = "#include <wine/test.h>\n\n";
+
+    for (auto& wlEntry : Whitelist)
+    {
+        string lineTableCode = CreateLinesTableCode(wlEntry);
+        code += lineTableCode;
+    }
+
+    code += "WHITELIST_FILE_ENTRY g_Whitelist[] =\n";
+    code += "{\n";
+    for (auto& wlEntry : Whitelist)
+    {
+        fprintf(stderr, "size of whitelist entry = %Iu\n", wlEntry.Lines.size());
+        // Get file name without extension
+        size_t nameLength = wlEntry.ShortFileName.rfind('.');
+        string sourceFileNameNoExt = wlEntry.ShortFileName.substr(0, nameLength);
+
+        string fileWhitelistName = "s_Whitelist_" + sourceFileNameNoExt;
+        code += "    { \"" + wlEntry.ShortFileName + "\", " + fileWhitelistName + ", _countof(" + fileWhitelistName + ")},\n";
+    }
+
+    code += "};\n\n";
+    code += "int g_WhitelistLength = _countof(g_Whitelist);\n";
+
+    return code;
+}
+
+int
+GenerateTableCode(string &WhitelistFileName, string &OutputFileName)
+{
+    vector<WL_ENTRY> whitelist = ReadWhitelist(WhitelistFileName);
+    string code = GenerateWhitelistCode(whitelist);
+    ofstream outputFile(OutputFileName);
+    outputFile << code;
+    return 0;
+}
+
+void Usage(void)
+{
+    printf("Usage: whitelister --add-entry <whitelist-file> <source-file> <line>\n");
+    printf("Usage: whitelister --gen-tables <whitelist-file> <output-file> <reactos source root>\n");
+}
+
+
+int main(int argc, const char* argv[])
+{
+    char oldpath[_MAX_PATH];
+    int result;
+
+    if (argc < 4)
+    {
+        Usage();
+        return 1;
+    }
+
+    if (string(argv[1]) == "--add-entry")
+    {
+        if (argc != 5)
+        {
+            printf("Usage: whitelister --add-entry <whitelist-file> <source-file> <line>\n");
```

Comment:
use `cout` ?

Reply:
iostream << operator performs so badly that it is almost always a bad idea to use it.
https://stackoverflow.com/questions/2872543/printf-vs-cout-in-c

Reply:
Yeah, I avoid it where I can as well. Only used it in one place for writing to a file.

Reply:
thanks for the info.